Remove "MLS" prefix on structs (#839)
mlswg/mls-protocol: draft-ietf-mls-protocol.md

    opaque authenticated_data<V>;

    ContentType content_type;

>     select (GroupContent.content_type) {

Reviewer: I suggested "MessageContent", "MessageContentTBS", etc. as opposed to using the "Group-" prefix. It got some 👍🏻, but no other reaction. Do you still prefer the "Group-" prefix?

>     select (GroupContent.content_type) {

Reviewer 2: ```suggestion
    select (FrameContent.content_type) {
```

        case application:
          opaque application_data<V>;
        case proposal:
          Proposal proposal;
        case commit:
          Commit commit;
    };

> } GroupContent;

Reviewer 2: ```suggestion
} FrameContent;
```

~~~ tls
struct {
    WireFormat wire_format;

>     GroupContent content;
>     GroupContentAuthData auth;

Reviewer 2: ```suggestion
    FrameContent content;
    FrameContentAuthData auth;
```

                                 +--------------+--------------+
                                                |
                                                V

>                                          GroupContent

Reviewer 2: ```suggestion
                                         FrameContent
```

                                             |  |                -.
                                             |  |                  |
                                    +--------+  |                  |
                                    |           |                  |
                                    V           |                  +-- Asymmetric

>                            GroupContentAuthData |                  |   Sign / Verify

Reviewer 2: ```suggestion
                           FrameContentAuthData |                  |   Sign / Verify
```

## Content Authentication

> GroupContent is authenticated using the GroupContentAuthData structure.

Reviewer 2: ```suggestion
FrameContent is authenticated using the FrameContentAuthData structure.
```

~~~ tls
struct {
    ProtocolVersion version = mls10;
    WireFormat wire_format;

>     GroupContent content;
>     select (GroupContentTBS.content.sender.sender_type) {

Reviewer 2: ```suggestion
    FrameContent content;
    select (FrameContentTBS.content.sender.sender_type) {
```

        case member:
        case new_member_commit:
            GroupContext context;
        case external:
        case new_member_proposal:
            struct{};
    };

> } GroupContentTBS;

Reviewer 2: ```suggestion
} FrameContentTBS;
```

opaque MAC<V>;

struct {

>     /* SignWithLabel(., "GroupContentTBS", GroupContentTBS) */

Reviewer 2: ```suggestion
    /* SignWithLabel(., "FrameContentTBS", FrameContentTBS) */
```

    opaque signature<V>;

>     select (GroupContent.content_type) {

Reviewer 2: ```suggestion
    select (FrameContent.content_type) {
```

        case proposal:
            struct{};
    };

> } GroupContentAuthData;

Reviewer 2: ```suggestion
} FrameContentAuthData;
```

~~~

The signature is computed using `SignWithLabel` with label

> `"GroupContentTBS"` and with a content that covers the message content and

Reviewer 2: ```suggestion
`"FrameContentTBS"` and with a content that covers the message content and
```

the `sender_type` of the sender as described above.

The confirmation tag value confirms that the members of the group have arrived

> at the same state of the group. A GroupContentAuthData is said to be valid when both

Reviewer 2: ```suggestion
at the same state of the group. A FrameContentAuthData is said to be valid when both
```

the `signature` and `confirmation_tag` fields are valid.

## Encoding and Decoding a Public Message

Messages that are authenticated but not encrypted are encoded using the PublicMessage structure.

~~~ tls
struct {

>     GroupContent content;
>     GroupContentAuthData auth;

Reviewer 2: ```suggestion
    FrameContent content;
    FrameContentAuthData auth;
```

    select (PublicMessage.content.sender.sender_type) {
        case member:
            MAC membership_tag;
        case external:
        case new_member_commit:
        case new_member_proposal:
            struct{};
    };
} PublicMessage;
~~~

The `membership_tag` field in the PublicMessage object authenticates the sender's
membership in the group. For messages sent by members, it MUST be set to the
following value:

~~~ tls
struct {
  GroupContentTBS content_tbs;
  GroupContentAuthData auth;
} AuthenticatedContentTBM;
~~~

~~~ pseudocode
membership_tag = MAC(membership_key, AuthenticatedContentTBM)
~~~

When decoding an PublicMessage into an AuthenticatedContent,
the application MUST check `membership_tag` and MUST check that the

> GroupContentAuthData is valid.

Reviewer 2: ```suggestion
FrameContentAuthData is valid.
```

          Commit commit;
    };

>     GroupContentAuthData auth;

Reviewer 2: ```suggestion
    FrameContentAuthData auth;
```

    uint64 epoch;
    ContentType content_type;
    opaque authenticated_data<V>;
} ContentAAD;
~~~

When decoding an AuthenticatedContentTBE, the application MUST check that the

> GroupContentAuthData is valid.

Reviewer 2: ```suggestion
FrameContentAuthData is valid.
```

field of the GroupInfo struct, while existing members can compute it directly.

Each Commit message updates these hashes by way of its enclosing
AuthenticatedContent.  The AuthenticatedContent struct is split into
ConfirmedTranscriptHashInput and InterimTranscriptHashInput. The former is used to
update the confirmed transcript hash and the latter to update the interim
transcript hash.

~~~ tls
struct {
    WireFormat wire_format;

>     GroupContent content; /* with content_type == commit */

Reviewer 2: ```suggestion
    FrameContent content; /* with content_type == commit */
```

## Proposals

> Proposals are included in a GroupContent by way of a Proposal structure

Reviewer 2: ```suggestion
Proposals are included in a FrameContent by way of a Proposal structure
```

} Proposal;
~~~

> On receiving a GroupContent containing a Proposal, a client MUST verify the

Reviewer 2: ```suggestion
On receiving a FrameContent containing a Proposal, a client MUST verify the
```

> signature inside GroupContentAuthData and that the `epoch` field of the enclosing
> GroupContent is equal to the `epoch` field of the current GroupContext object.

Reviewer 2: ```suggestion
signature inside FrameContentAuthData and that the `epoch` field of the enclosing
FrameContent is equal to the `epoch` field of the current GroupContext object.
```

  of PSKs in the derivation corresponds to the order of PreSharedKey proposals
  in the `proposals` vector.

> * Construct a GroupContent object containing the Commit object. Sign the
>   GroupContent using the old GroupContext as context.
>   * Use the GroupContent to update the confirmed transcript hash and update

Reviewer 2: ```suggestion
* Construct a FrameContent object containing the Commit object. Sign the
  FrameContent using the old GroupContext as context.
  * Use the FrameContent to update the confirmed transcript hash and update
```

  * Use the `confirmation_key` for the new epoch to compute the
    `confirmation_tag` value.
  * Calculate the interim transcript hash using the new confirmed transcript

>     hash and the `confirmation_tag` from the GroupContentAuthData.

Reviewer 2: ```suggestion
    hash and the `confirmation_tag` from the FrameContentAuthData.
```

* Protect the AuthenticatedContent object using keys from the old epoch:
  * If encoding as PublicMessage, compute the `membership_tag` value using the
    `membership_key`.
  * If encoding as an PrivateMessage, encrypt the message using the
    `sender_data_secret` and the next (key, nonce) pair from the sender's
    handshake ratchet.

* Construct a GroupInfo reflecting the new state:
  * Group ID, epoch, tree, confirmed transcript hash, interim transcript
    hash, and group context extensions from the new state

>   * The confirmation_tag from the GroupContentAuthData object

Reviewer 2: ```suggestion
  * The confirmation_tag from the FrameContentAuthData object
```

A member of the group applies a Commit message by taking the following steps:

> * Verify that the `epoch` field of the enclosing GroupContent is equal

Reviewer 2: ```suggestion
* Verify that the `epoch` field of the enclosing FrameContent is equal
```

  to the `epoch` field of the current GroupContext object

* Unprotect the Commit using the keys from the current epoch:
  * If the message is encoded as PublicMessage, verify the membership MAC using
    the `membership_key`
  * If the message is encoded as PrivateMessage, decrypt the message using the
    `sender_data_secret` and the (key, nonce) pair from the step on the sender's
    hash ratchet indicated by the `generation` field.

> * Verify that the signature on the GroupContent message as described in

Reviewer 2: ```suggestion
* Verify that the signature on the FrameContent message as described in
```

* Use the `confirmation_key` for the new epoch to compute the confirmation tag
  for this message, as described below, and verify that it is the same as the

>   `confirmation_tag` field in the GroupContentAuthData object.

Reviewer 2: ```suggestion
  `confirmation_tag` field in the FrameContentAuthData object.
```

# Sequencing of State Changes {#sequencing}

Each Commit message is premised on a given starting state,

> indicated by the `epoch` field of the enclosing GroupContent.

Reviewer 2: ```suggestion
indicated by the `epoch` field of the enclosing FrameContent.
```
